internal/controller/multiclusterservice_controller.go

```diff
@@ -29,10 +37,132 @@ type MultiClusterServiceReconciler struct {
+		return ctrl.Result{}, fmt.Errorf("failed to reconcile ClusterProfile: %w", err)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// HelmChartOpts returns slice of helm chart options to use with Sveltos.
+// Namespace is the namespace of the referred templates in services slice.
+func HelmChartOpts(ctx context.Context, c client.Client, l logr.Logger, namespace string, services []hmc.ServiceSpec) ([]sveltos.HelmChartOpts, error) {
+	opts := []sveltos.HelmChartOpts{}
+
+	// NOTE: The Profile/ClusterProfile object will be updated with
+	// no helm charts if len(mc.Spec.Services) == 0. This will result
+	// in the helm charts being uninstalled on matching clusters if
+	// Profile/ClusterProfile originally had len(m.Spec.Sevices) > 0.
+	for _, svc := range services {
+		if svc.Disable {
+			l.Info(fmt.Sprintf("Skip adding Template %s because Disable=true", svc.Template))
+			continue
+		}
+
+		tmpl := &hmc.ServiceTemplate{}
+		// Here we can use the same namespace for all services
+		// because if the services slice is part of:
+		// 1. ManagedCluster: Then the referred template must be in its own namespace.
+		// 2. MultiClusterService: Then the referred template must be in hmc-system namespace.
+		ref := types.NamespacedName{Name: svc.Template, Namespace: namespace}
+		if err := c.Get(ctx, ref, tmpl); err != nil {
+			return nil, fmt.Errorf("failed to get Template %s: %w", ref.String(), err)
+		}
+
+		source, err := TemplateSource(ctx, c, tmpl)
+		if err != nil {
+			return nil, fmt.Errorf("could not get repository url: %w", err)
+		}
+
+		opts = append(opts, sveltos.HelmChartOpts{
+			Values:        svc.Values,
+			RepositoryURL: source.Spec.URL,
+			// We don't have repository name so chart name becomes repository name.
+			RepositoryName: tmpl.Spec.Helm.ChartName,
+			ChartName: func() string {
+				if source.Spec.Type == utils.RegistryTypeOCI {
+					return tmpl.Spec.Helm.ChartName
+				}
+				// Sveltos accepts ChartName in <repository>/<chart> format for non-OCI.
+				// We don't have a repository name, so we can use <chart>/<chart> instead.
+				// See: https://projectsveltos.github.io/sveltos/addons/helm_charts/.
+				return fmt.Sprintf("%s/%s", tmpl.Spec.Helm.ChartName, tmpl.Spec.Helm.ChartName)
+			}(),
+			ChartVersion: tmpl.Spec.Helm.ChartVersion,
+			ReleaseName:  svc.Name,
+			ReleaseNamespace: func() string {
+				if svc.Namespace != "" {
+					return svc.Namespace
+				}
+				return svc.Name
+			}(),
+			// The reason it is passed to PlainHTTP instead of InsecureSkipTLSVerify is because
+			// the source.Spec.Insecure field is meant to be used for connecting to repositories
+			// over plain HTTP, which is different than what InsecureSkipTLSVerify is meant for.
+			// See: https://github.com/fluxcd/source-controller/pull/1288
+			PlainHTTP: source.Spec.Insecure,
+		})
+	}
+
+	return opts, nil
+}
+
+func (r *MultiClusterServiceReconciler) reconcileDelete(ctx context.Context, mcsvc *hmc.MultiClusterService) (ctrl.Result, error) {
```

Comment:
The reason is so that Sveltos can clean up the resources that depend on `Profile` or `ClusterProfile` object. See https://github.com/Mirantis/hmc/pull/362#discussion_r1775019326 for previous discussion.

Reply:
Please correct me if I'm mistaken: **without** the current logic and relying only on the owner references, removing an `mcs` object will remove the related `ClusterProfile` object, so in this way, Sveltos **will NOT** garbage collect other objects relying on that cluster profile?

Reply:
If deleting the parent will set the `DeletionTimestamp` on the child objects, then we might not need to explicitly delete the child object because Sveltos checks `DeletionTimestamp`. I don't know the answer to that. I can try and test it.

Reply:
@zerospiel You are correct in that we can rely on owner references for removing dependent objects. However, in this case we need to issue a delete on `Profile`. The reason is that without this [`DeleteProfile`](https://github.com/Mirantis/hmc/blob/3f9a2addc4b5f9feda254a7d94afa594a2e9ae7b/internal/controller/managedcluster_controller.go#L591-L594) call, the Sveltos addon-controller keeps on looping with the following error:
```sh

I1016 23:21:22.030322       1 controller.go:302] "Reconciling" controller="clustersummary" controllerGroup="config.projectsveltos.io" controllerKind="ClusterSummary" ClusterSummary="hmc-system/p--wali-aws-dev-capi-wali-aws-dev" namespace="hmc-system" name="p--wali-aws-dev-capi-wali-aws-dev" reconcileID="3d98ee4c-2713-45bf-ae9b-fa7f7000fd77"
I1016 23:21:22.030397       1 clustersummary_controller.go:122] "Reconciling" controller="clustersummary" controllerGroup="config.projectsveltos.io" controllerKind="ClusterSummary" ClusterSummary="hmc-system/p--wali-aws-dev-capi-wali-aws-dev" namespace="hmc-system" name="p--wali-aws-dev-capi-wali-aws-dev" reconcileID="3d98ee4c-2713-45bf-ae9b-fa7f7000fd77"
I1016 23:21:22.030893       1 clustersummary_controller.go:225] "Reconciling ClusterSummary delete" controller="clustersummary" controllerGroup="config.projectsveltos.io" controllerKind="ClusterSummary" ClusterSummary="hmc-system/p--wali-aws-dev-capi-wali-aws-dev" namespace="hmc-system" name="p--wali-aws-dev-capi-wali-aws-dev" reconcileID="3d98ee4c-2713-45bf-ae9b-fa7f7000fd77"
E1016 23:21:22.045361       1 clustersummary_controller.go:250] "failed to remove ResourceSummary." err="failed to get API group resources: unable to retrieve the complete list of server APIs: apiextensions.k8s.io/v1: Get \"https://wali-aws-dev-apiserver-1185732421.ca-central-1.elb.amazonaws.com:6443/apis/apiextensions.k8s.io/v1\": dial tcp: lookup wali-aws-dev-apiserver-1185732421.ca-central-1.elb.amazonaws.com on 10.96.0.10:53: no such host" controller="clustersummary" controllerGroup="config.projectsveltos.io" controllerKind="ClusterSummary" ClusterSummary="hmc-system/p--wali-aws-dev-capi-wali-aws-dev" namespace="hmc-system" name="p--wali-aws-dev-capi-wali-aws-dev" reconcileID="3d98ee4c-2713-45bf-ae9b-fa7f7000fd77"
I1016 23:21:22.045893       1 controller.go:318] "Reconcile done, requeueing after 10s" controller="clustersummary" controllerGroup="config.projectsveltos.io" controllerKind="ClusterSummary" ClusterSummary="hmc-system/p--wali-aws-dev-capi-wali-aws-dev" namespace="hmc-system" name="p--wali-aws-dev-capi-wali-aws-dev" reconcileID="3d98ee4c-2713-45bf-ae9b-fa7f7000fd77"
```
So during cleanup, it tried to access the deleted target cluster and fails. This might actually be a bug in Sveltos addon-controller.

Reply:
Thanks for the check, the error looks unrelated though (I'm not much in the details of the sveltos). I've crawled through the addon-ctrl code and that particular error should not be related to my understanding (`*Profile` creation leads to `ClusterSummary` creation, and deletion of the `*Profile` either way leads to the deletion of the `ClusterSummary`, I've found no diff between having explicit/implicit profiles deletion regarding this behavior). Potential diff may be related to the cluster (either CAPI or Sveltos) readiness but I have no child clusters on-hands so cannot check the actual states, anyway, I do not think that is related to the problem.

In conclusion: if you are sure that the setting finalizer somehow changes the states of a cluster and depending on that state the sveltos' clustersummary controller behaves in different ways, then it's okay. 

I'd rather left comments here (`Delete(Cluster)Profile` functions) why do we even need them